### DISCUSSION TITLE
Can GoTrue's currentSessionOrNull be used for a multi-threaded Kotlin server?

### DISCUSSION THREAD

**Discussion by franzvezuli** [supabase-community/supabase-kt#396] [General]
Hello,

I'm new to Supabase, but want to use it for a project. I am using Kotlin as a server (not Android), to be specific its a https://ktor.io application.

I wrote a test that invokes my apps registration endpoint in parallel `n times` and noticed that it does not retrieve the correct session thats related to the user that just registered, which leads me to believed this is supposed to be used only for a single user login context? (which for an Android app makes sense).

I'm instantiating the client like so
```
val client = createSupabaseClient( ... ) {
  install(Auth)
}
```

When using something like this
```
client.auth.signUpWith(Email) {
  email = ...
  password = ...
}
```
then invoke
```
val user = client.auth.currentSessionOrNull()?.user!!
```

Should I be using Supabase REST API instead of this library for multi-threaded applications maybe? Or am I just using it incorrectly?

I also know `signUpWith` returns the user immediately, which I can avoid currentSessionOrNull(), but:

1. It does not contain the accessToken, which I need to pass to my front end. I would need to do a separate query by id or email, and then also figure out how to get a supabase accessToken (or mint my own, which is not ideal).
2. The sister method of `client.auth.signInWith(Email)` does not seem to return anything at all, not even a user.

Thanks! Let me know.

**Comment by jan-tennert** [comment]
Hey, its definitely possible to use this library for such a use-case. The `signUpWith` method only returns a partial user, which has to get verified. You can also enable auto-confirm which makes this method do the exact thing as `signInWith`, which saves the session in memory (and normally also somewhere persistent). You can get the session like you pointed out (also with `currentUserOrNull`. You might want to check https://github.com/supabase-community/supabase-kt/issues/366, where this library is getting used similary.

**Reply by franzvezuli** [reply]
Thank you so much for the reply! I'm reading over those `install(Auth) { // settings }` now and they seem like they might do the trick.

Thanks again!
